cloudinit/distros/package_management/apt.py

```diff
@@ -0,0 +1,213 @@
+    """Parse the 'apt_get_wrapper' section of cloud-config.
+
+    apt_get_wrapper may be defined in cloud-config:
+      apt_get_wrapper:
+        enabled: true
+        command: ["eatmydata"]
+
+    The function takes the value of "apt_get_wrapper" and returns the list
+    of arguments to prefix to the apt-get command.
+    """
+    if not cfg:
+        enabled = "auto"
+        command = ["eatmydata"]
+    else:
+        enabled = cfg.get("enabled")
+        command = cfg.get("command")
+
+        if isinstance(command, str):
+            command = [command]
+        elif not isinstance(command, list):
+            raise TypeError("apt_wrapper command must be a string or list")
+
+    if util.is_true(enabled) or (
+        str(enabled).lower() == "auto" and command and subp.which(command[0])
+    ):
+        return command
+    else:
+        return []
+
+
+class Apt(PackageManager):
+    name = "apt"
+
+    def __init__(
+        self,
+        runner: helpers.Runners,
+        *,
+        apt_get_wrapper_command: Sequence[str] = (),
+        apt_get_command: Optional[Sequence[str]] = None,
+        apt_get_upgrade_subcommand: Optional[str] = None,
+    ):
+        super().__init__(runner)
+        if apt_get_command is None:
+            apt_get_command = APT_GET_COMMAND
+        if apt_get_upgrade_subcommand is None:
+            apt_get_upgrade_subcommand = "dist-upgrade"
+        self.apt_command = tuple(apt_get_wrapper_command) + tuple(apt_get_command)
+
+        self.apt_get_upgrade_subcommand = apt_get_upgrade_subcommand
+        self.environment = os.environ.copy()
+        self.environment["DEBIAN_FRONTEND"] = "noninteractive"
+
+    @classmethod
+    def from_config(cls, runner: helpers.Runners, cfg: Mapping) -> "Apt":
+        return Apt(
+            runner,
+            apt_get_wrapper_command=get_apt_wrapper(cfg.get("apt_get_wrapper")),
+            apt_get_command=cfg.get("apt_get_command"),
+            apt_get_upgrade_subcommand=cfg.get("apt_get_upgrade_subcommand"),
+        )
+
+    def update_package_sources(self):
+        self.runner.run(
+            "update-sources",
+            self.run_package_command,
+            ["update"],
+            freq=PER_INSTANCE,
+        )
+
+    @functools.lru_cache(maxsize=1)
+    def get_all_packages(self):
+        resp: str = subp.subp(["apt-cache", "pkgnames"]).stdout
+
+        # Searching the string directly and searching a list are both
+        # linear searches. Converting to a set takes some extra up front
+        # time, but resulting searches become binary searches and are much
+        # faster
+        return set(resp.splitlines())
+
+    def get_unavailable_packages(self, pkglist: Iterable[str]):
+        return [pkg for pkg in pkglist if pkg not in self.get_all_packages()]
+
+    def install_packages(self, pkglist: Iterable[str]) -> UninstalledPackages:
+        self.update_package_sources()
+        unavailable = self.get_unavailable_packages(pkglist)
```

Comment:
While not thrilled about the cost of this lookup function I get that it improves our error handling and fallback to other package managers...  Given that apt-get install X Y NOT_PRESENT doesn't give us the ability to optionally install X and Y if NOT_PRESENT isn't available, this is the best option to look before we leap.